experimental/sso-with-skills/RootBot/TokenExchangeSkillHandler.cs

```diff
@@ -0,0 +1,161 @@
+                            {
+                                context.TurnState.Add<IIdentity>("BotIdentity", claimsIdentity);
+
+                                // AAD token exchange
+                                try
+                                {
+                                    var result = await _tokenExchangeProvider.ExchangeTokenAsync(
+                                        context,
+                                        _connectionName,
+                                        activity.Recipient.Id,
+                                        new TokenExchangeRequest() { Uri = oauthCard.TokenExchangeResource.Uri }).ConfigureAwait(false);
+
+                                    if (!string.IsNullOrEmpty(result.Token))
+                                    {
+                                        // Send an Invoke back to the Skill
+                                        return await SendTokenExchangeInvokeToSkill(activity, oauthCard.TokenExchangeResource.Id, result.Token, oauthCard.ConnectionName, targetSkill, default(CancellationToken)).ConfigureAwait(false);
+                                    }
+                                }
+                                catch
+                                {
+                                    // Show oauth card if token exchange fails.
+                                    return false;
+                                }
+
+                                return false;
+                            }
+                        }
+                    }
+                }
+            }
+            return false;
+        }
+
+        private async Task<bool> SendTokenExchangeInvokeToSkill(Activity incomingActivity, string id, string token, string connectionName, BotFrameworkSkill targetSkill, CancellationToken cancellationToken)
+        {
+            var activity = incomingActivity.CreateReply() as Activity;
+            activity.Type = ActivityTypes.Invoke;
+            activity.Name = SignInConstants.TokenExchangeOperationName;
+            activity.Value = new TokenExchangeInvokeRequest()
+            {
+                Id = id,
+                Token = token,
+                ConnectionName = connectionName,
+            };
+
+            var skillConversationReference = await _conversationIdFactory.GetSkillConversationReferenceAsync(incomingActivity.Conversation.Id, cancellationToken).ConfigureAwait(false);
+            activity.Conversation = skillConversationReference.ConversationReference.Conversation;
+            activity.ServiceUrl = skillConversationReference.ConversationReference.ServiceUrl;
+
+            // route the activity to the skill
+            var response = await _skillClient.PostActivityAsync(_botId, targetSkill, _skillsConfig.SkillHostEndpoint, activity, cancellationToken);
+
+            // Check response status: true if success, false if failure
+            return (response.Status >= 200 && response.Status <= 299);
```

Comment:
HttpStatusCodes.OK, etc

Reply:
will use IsSuccessCode from InvokeResponse once R9 is released